feat: reduce build log verbosity

eduardoboucas opened the pull request:
#### Summary

⏳

eduardoboucas commented on packages/build/src/log/messages/core.ts:
```diff
@@ -32,10 +33,18 @@ export const logBuildSuccess = function (logs) {
 }
 
-export const logTimer = function (logs, durationNs, timerName, systemLog) {
+export const logTimer = function (logs, prefix, durationNs, timerName, systemLog, outputManager?: OutputManager) {
```
This additional parameter lets us print an empty line before the timer, which was the sole purpose of the `logStepSuccess` function we can now remove.

eduardoboucas commented on packages/build/src/log/messages/install.js:
```diff
@@ -1,20 +1,13 @@
 import { log, logArray, logSubHeader } from '../logger.js'
 
-export const logInstallMissingPlugins = function (logs, packages) {
-  const runtimes = packages.filter((pkg) => isRuntime(pkg))
```
This was always empty, because `isRuntime` expects an object with a `packageName` property whereas the `packages` parameter was an array of strings. I've changed the function to receive both.

eduardoboucas commented on packages/build/src/steps/run_step.ts:
```diff
@@ -114,8 +115,14 @@ export const runStep = async function ({
+
+    let outputManager: OutputManager | undefined
+
+    if (coreStepId === undefined && featureFlags.netlify_build_reduced_output) {
```
We don't want to use the output manager for core steps, since those run in the same process as Netlify Build and so piping stdout/stderr is not that simple. We can better control the output of core steps though, so that's okay.

eduardoboucas commented:
> I think a test for the new behaviour would be great.

The challenge is that the feature flag will affect a ton of tests, and there's no easy way to make a flag enabled by default for all tests without enabling it by default in production. So, as it stands, the tests are just asserting that there's no regression with the flag disabled.

What I'll do is push a commit with the flag enabled and the snapshots updated, and once that passes we can see what the changes look like and verify that the tests are passing. We can then revert the commit and merge with the flag disabled. Does that sound okay?

> Also, I found the name `OutputGate` confusing. In my understanding, a `Gate` is something that can block data from going through, and `opening` the gate is what allows that to happen. Just by the names, it sounds like it's intended to buffer up plugin output until a certain condition is met.

That's exactly what it does. By default, plugin output will not go through unless there's a code path that opens the gate. That happens when the plugin emits something to stderr or stdout. So I think the term "gate" makes sense here.

> Also, this change makes it a lot harder to understand which plugins were being run, also for community plugins. Does it make sense to only roll this out for the next-runtime in the beginning, and then see how that goes?

Not sure I agree here. We'll always print the name and source of each plugin being loaded: https://github.com/netlify/build/blob/8982903b7724d736c9f9514aed8f6c960c6378ba/packages/build/src/log/messages/compatibility.ts#L45

Also, the `--debug` flag will still show the full output, so that can be used for troubleshooting if needed. I think the main goal here is to hide things that don't provide useful information, and I think an empty plugin event block is an example of that.

eduardoboucas commented:
> I still think we should have at least one test that has this flag enabled. That way, we have some coverage for `OutputGate`, even after merging this. In the current state, that entire class isn't covered by any test.
> 
> >

Sounds good. I'll do that after reverting the latest commit.

Skn0tt commented:
I think the difference between our mental models is that I think of the plugin output as what's being gated, and you think of the plugin header as what's being gated. Because "gate" is closely related to "stream" in my mind, and the plugin output tends to be a stream whereas the header is not, I couldn't imagine what you meant and got confused! I think the rename should clear things up, yes!

eduardoboucas commented:
@Skn0tt okay, so I pushed a few commits:

- https://github.com/netlify/build/pull/5643/commits/bbf7374f0e1e2d0a91315e9108a59b1832297eba: You can see how the snapshots will get changed with the flag enabled
- https://github.com/netlify/build/pull/5643/commits/102d1c904d8607ca5c065fe9f07f0380bf161bc8: You can see tests passing
- https://github.com/netlify/build/pull/5643/commits/0f964b3f8171e7c90e32f745ea94d6b4a51da40a: I've renamed `OutputGate` to `OutputFlusher` — hopefully that makes more sense to you?
- https://github.com/netlify/build/pull/5643/commits/1606455e673c46b90870d43d25e828ce630e19f1: I've added the test you requested

Skn0tt reviewed (approved):
Wonderful, thanks for addressing my feedback!

Skn0tt commented:
I'm including this PR in my prayers for a speedy GitHub Actions recovery 🙏
